scanner/scanner.go

```diff
@@ -0,0 +1,191 @@
+	// index of nasty files.
+	sl.Skylink = ""
+	sl.Status = database.SkylinkStatusComplete
+	sl.Infected = inf
+	sl.InfectionDescription = desc
+	sl.Size = size
+	sl.ScannedAllContent = scannedSize == size
+	sl.ScannedAllOffsets = false
+	sl.Timestamp = time.Now().UTC()
+	err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+	if err != nil {
+		s.staticLogger.Debugln(errors.AddContext(err, "updating a skylink's status failed"))
+	}
+	return err
+}
+
+// Start launches a background task that periodically scans the database for
+// new skylink records and sends them for scanning.
+func (s Scanner) Start() {
+	// abort channel which interrupts the current scanning operation
+	abort := make(chan bool)
+
+	// Start a thread that watches the context and immediately closes the abort
+	// channel when the context is closed. This will immediately (or at least as
+	// quickly as ClamAV allows) terminate the current scan and allow for a
+	// quick and clean service shutdown.
+	go func() {
+		<-s.staticCtx.Done()
+		close(abort)
+	}()
+
+	// Start the scanning loop.
+	go func() {
+		// sleepLength defines how long the thread will sleep before scanning
+		// the next skylink. Its value is controlled by SweepAndScan - while we
+		// keep finding files to scan, we'll keep this sleep at zero. Once we
+		// run out of files to scan we'll reset it to its full duration of
+		// sleepBetweenScans.
+		sleepLength := sleepBetweenScans
+		for {
+			numSubsequentErrs := 0
+			select {
+			case <-s.staticCtx.Done():
+				return
+			case <-time.After(sleepLength):
+				err := s.SweepAndScan(abort)
+				if errors.Contains(err, database.ErrNoDocumentsFound) {
+					// This was a successful call, so the number of subsequent
+					// errors is reset and we sleep for a pre-determined period
+					// in waiting for new skylinks to be uploaded.
+					sleepLength = sleepBetweenScans
+					numSubsequentErrs = 0
+				} else if err != nil {
+					// On error, we sleep for an increasing amount of time -
+					// from 100ms on the first error to 100s on the fourth and
+					// subsequent errors.
+					sleepLength = sleepOnErrStep * time.Duration(math.Pow10(numSubsequentErrs))
+					numSubsequentErrs++
+					if numSubsequentErrs > sleepOnErrSteps {
+						numSubsequentErrs = sleepOnErrSteps
+					}
+				} else {
+					// A successful scan. Reset the number of subsequent errors.
+					numSubsequentErrs = 0
+					// No need to sleep after a successful scan.
+					sleepLength = 0
+				}
+			}
+		}
+	}()
+}
+
+// StartUnlocker launches a background thread that periodically scans the
+// database and resets the state of potentially stuck scans. If a scan has been
+// initiated too long ago it will put it back in "new" state, so it can be
+// retried.
+func (s Scanner) StartUnlocker() {
+	go func() {
+		ticker := time.NewTicker(database.ScanTimeout)
+		for {
+			select {
+			case <-s.staticCtx.Done():
+				return
+			case <-ticker.C:
+				n, err := s.staticDB.CancelStuckScans(s.staticCtx)
```

Comment:
f/u move this out of the `case <-ticker.C` to reduce the x-wing factor of this method.

Reply:
Will do in #2